src/SignalFx.LambdaWrapper/TelemetryTracker.cs

```diff
@@ -0,0 +1,114 @@
+﻿using System;
+using System.Collections.Generic;
+
+using Amazon.Lambda.Core;
+using SignalFx.Tracing;
+using SignalFx.Tracing.Headers;
+
+namespace SignalFx.LambdaWrapper
+{
+    /// <summary>
+    /// This is a helper that is used to wrap AWS Lambda functions with telemetry data.
+    /// </summary>
+    internal struct TelemetryTracker : IDisposable
+    {
+        internal static readonly Tracer s_sfxTracer;
+        internal static bool s_coldStart;
+
+        internal readonly Scope _sfxScope;
+        internal readonly MetricWrapper _metricWrapper;
+
+        static TelemetryTracker()
+        {
+            s_coldStart = true;
+            if (TelemetryConfiguration.TracingEnabled)
+            {
+                // This forces the registration of the SFx tracer with OpenTracing.
+                s_sfxTracer = Tracer.Instance;
+            }
+        }
+
+        internal TelemetryTracker(
+            ILambdaContext lambdaContext,
+            string operationName = null,
+            IEnumerable<KeyValuePair<string, string>> tags = null,
+            IHeadersCollection headersCollection = null)
+            : this()
+        {
+            if (!TelemetryConfiguration.TracingEnabled && !TelemetryConfiguration.MetricsEnabled)
+            {
+                return;
+            }
+
+            var coldStart = s_coldStart;
+            if (coldStart)
+            {
+                s_coldStart = false;
+            }
+
+            var commonTags = lambdaContext.ExtractCommonTags();
+
+            if (TelemetryConfiguration.MetricsEnabled)
+            {
+                _metricWrapper = new MetricWrapper(commonTags);
+            }
+
+            if (TelemetryConfiguration.TracingEnabled)
+            {
+                ISpanContext parentContext = null;
+                if (TelemetryConfiguration.ContextPropagationEnabled && headersCollection != null)
+                {
+                    parentContext = B3SpanContextPropagator.Instance.Extract(headersCollection);
```

Comment:
Would we want to use the `Tracer.Extract` and let its configuration determine if B3 is registered for headers?

Reply:
That is a good idea and more forward-looking. I was focused on avoiding the OT tracer to avoid the allocation of wrappers that surround the concrete tracker. That said I will take a look to see if the change gets too big for this stage.

Reply:
Done.